src/cpp/src/visual_language/nanollava/classes.cpp

```diff
@@ -0,0 +1,289 @@
+        return encoded_input_ids;
+    }
+}
+
+ov::Tensor insert_image_placeholders(const ov::Tensor& tokens, size_t image_placeholder_size) {
+    size_t tokens_size = tokens.get_size();
+    const int64_t* tokens_data = tokens.data<const int64_t>();
+    size_t images_num = 0;
+    for (size_t idx = 0; idx < tokens_size; idx++) {
+        if (tokens_data[idx] == IMAGE_PLACEHOLDER)
+            images_num++;
+    }
+    size_t new_tokens_size = tokens_size + images_num * image_placeholder_size - images_num;
+    ov::Tensor res_tokens(tokens.get_element_type(), {1, new_tokens_size});
+
+    int64_t* res_tokens_data = res_tokens.data<int64_t>();
+
+    size_t idx = 0;
+    size_t token_idx = 0;
+    while (idx < new_tokens_size) {
+        if (tokens_data[token_idx] == IMAGE_PLACEHOLDER) {
+            for (size_t i = 0; i < image_placeholder_size; i++) {
+                res_tokens_data[idx + i] = IMAGE_PLACEHOLDER;
+            }
+            idx += image_placeholder_size;
+        }
+        else {
+            res_tokens_data[idx] = tokens_data[token_idx];
+            idx++;
+        }
+        token_idx++;
+    }
+    return res_tokens;
+}
+
+ov::Tensor drop_image_placeholders(const ov::Tensor& tokens, size_t image_placeholder_size) {
+    size_t tokens_size = tokens.get_size();
+    const int64_t* tokens_data = tokens.data<const int64_t>();
+    size_t images_num = 0;
+    for (size_t idx = 0; idx < tokens_size; idx++) {
+        if (tokens_data[idx] == IMAGE_PLACEHOLDER)
+            images_num++;
+    }
+    images_num = images_num / image_placeholder_size;
+
+    size_t new_tokens_size = tokens_size - images_num * image_placeholder_size + images_num;
+    ov::Tensor res_tokens(tokens.get_element_type(), {1, new_tokens_size});
+
+    int64_t* res_tokens_data = res_tokens.data<int64_t>();
+
+    size_t idx = 0;
+    size_t res_idx = 0;
+    while (idx < tokens_size) {
+        if (tokens_data[idx] == IMAGE_PLACEHOLDER) {
+            idx += image_placeholder_size;
+            res_tokens_data[res_idx] = IMAGE_PLACEHOLDER;
+        }
+        else {
+            res_tokens_data[res_idx] = tokens_data[idx];
+            idx++;
+        }
+        res_idx++;
+    }
+    return res_tokens;
+}
+
+ov::Tensor InputsEmbedderNanoLLaVA::get_encoded_input_ids(const std::string& prompt, ov::genai::VLMPerfMetrics& metrics) {
+    auto new_chat_tokens = apply_chat_template_tokenize(prompt, metrics);
+
+    if (m_image_features_size > 0)
+        new_chat_tokens = insert_image_placeholders(new_chat_tokens, m_image_features_size);
+    auto new_input_ids = IInputsEmbedder::update_history(new_chat_tokens);
+    m_prev_hist_length = m_kv_cache_state.get_state().size();
+    m_kv_cache_state.add_inputs(new_input_ids);
+
+    if (m_image_features_size > 0)
+        new_input_ids = drop_image_placeholders(new_input_ids, m_image_features_size);
+    return new_input_ids;
+}
```

Comment:
Is it possible to insert image placeholders (`m_image_features_size` times) in `apply_chat_template_tokenize(...)` or in `tokenize_without_image_tag(...)`?
If so, we will not need to change default `get_encoded_input_ids(...)` method.
And if we need to remove image placeholders later then, it can be done in `InputsEmbedderNanoLLaVA::get_inputs_embeds()`, after calling `get_encoded_input_ids()` before passing `input_ids` to text embedding model.